### PR TITLE
Add Dart Windows support

### PR DESCRIPTION
**Pull request by mit-mit**
As discussed in https://travis-ci.community/t/dart-support-on-windows/450/2

cc @BanzaiMan can we give this a run in staging?

### CODE REVIEW

**Comment by BanzaiMan** [comment]
This has been deployed to https://staging.travis-ci.com.

**Comment by BanzaiMan** [comment]
Oops. Sorry about that. I've enabled `dart` and `r` just now.

**Comment by mit-mit** [comment]
@BanzaiMan I updated the PR in an attempt to resolve the issue with the `.bat` files not being in path. Can we deploy this to staging?

**Comment by BanzaiMan** [comment]
@mit-mit This has been deployed to https://staging.travis-ci.com.

**Comment by BanzaiMan** [comment]
This has been deployed to https://staging.travis-ci.com again.

**Comment by mit-mit** [comment]
@BanzaiMan would mind deploying this again to staging?

**Comment by BanzaiMan** [comment]
@mit-mit This has been deployed to https://staging.travis-ci.com/

**Comment by mit-mit** [comment]
@BanzaiMan I'm having lots of issues on staging where trying to view the log errors out with 'There was an error while trying to fetch the log.' ?

**Comment by BanzaiMan** [comment]
This has been deployed to https://staging.travis-ci.com/. I apologize in advance that we may have to be deployed over without notice.

**Comment by mit-mit** [comment]
@BanzaiMan I'm consistently having issues with viewing the logs from any Windows build:

screenshot 2018-12-19 at 11 36 35

**Comment by mit-mit** [comment]
@BanzaiMan does https://staging.travis-ci.com still have this PR applied? I'd like to pick work on this back up again.

**Comment by BanzaiMan** [comment]
@mit-mit Sure thing! I've deployed this to https://staging.travis-ci.com. We'll try to leave it there as long as we can.

**Comment by mit-mit** [comment]
cc @DanTup for review

**Comment by DanTup** [comment]
The alias changes seem reasonable to me, there's already one in the file, suggesting that it should work:

> sh.raw 'function dartfmt() { pub run dart_style:format "$@"; }'

(note: you're redefining `dartfmt` so possible this one can be removed, assuming it works).

Do we need to deploy this to staging for testing? There's some info in the readme about running it in docker - if this doesn't fix it and we should be able to run locally, I don't mind having a go to see to reduce the time for debugging.

**Comment by mit-mit** [comment]
@BanzaiMan can we please try to deploy this to staging?

**Comment by BanzaiMan** [comment]
@mit-mit This has been deployed to com-staging.

**Comment by mit-mit** [comment]
@BanzaiMan sorry, we had a dumb error in the last version, can we try to deploy again?

**Comment by DanTup** [comment]
Also - I made some progress running this locally that may help, but when I run `travis compile` I get this error:

```
E RuntimeError: data.slug must not be empty
/Users/dantup/.travis/travis-build/lib/travis/build/data.rb:154:in `slug'
/Users/dantup/.travis/travis-build/lib/travis/build/appliances/wait_for_network.rb:20:in `block in check_urls'
```

I found some other instances of this around, but they all appear to have been fixed. It's a fresh clone of this repo today so I believe it's current, and with the correct dependency versions. Can you advise where that should come from? I'm running it in a folder that has this `.travis.yml` file.

```yaml
language: dart
os:
  - windows
  - linux
  - osx
sudo: false

dart_task:
  dartfmt: true
```

**Comment by mit-mit** [comment]
OK, all fixes we believe we need. @BanzaiMan we'd very much appreciate a deploy to staging

**Comment by BanzaiMan** [comment]
Awesome. This has been deployed to .com staging.

**Comment by mit-mit** [comment]
This appears to be working now :-). cc @nex3 @a14n can you review, please?

File: lib/travis/build/script/dart.rb [hunk 1]
```diff
@@ -92,13 +92,26 @@ def setup
...
+            if os == 'windows'
+              # Define commands; on Windows git bash requires .bat extensions
+              # https://github.com/msysgit/msysgit/issues/101
+              sh.raw 'function dart2js() { dart2js.bat "$@"; }'
+              sh.raw 'function dartanalyzer() { dartanalyzer.bat "$@"; }'
+              sh.raw 'function dartdevc() { dartdevc.bat "$@"; }'
+              sh.raw 'function dartdevk() { dartdevk.bat "$@"; }'
+              sh.raw 'function dartdoc() { dartdoc.bat "$@"; }'
+              sh.raw 'function dartfmt() { dartfmt.bat "$@"; }'
+              sh.raw 'function pub() { pub.bat "$@"; }'
```
Comment (a14n) [inline lib/travis/build/script/dart.rb]: Is `dart` working differently? Should a `dart` command be also defined?

Comment (DanTup) [reply lib/travis/build/script/dart.rb]: I think the issue is only with `.bat` files inside the Bash shell on Windows. I think `.exe`'s are selected correctly in preference to the extensionless files.

**Comment by mit-mit** [comment]
Thanks for the review @a14n !

@BanzaiMan what the process for getting this merged?

**Comment by kevmoo** [comment]
Friendly ping, @BanzaiMan 😄

**Comment by BanzaiMan** [comment]
Hi, everyone, sorry for the delay. We have a bit of back log on this repo at the moment that we hope to resolve in the very near future. Please accept our apologies for the delay.

In the mean time, this has been deployed to .com staging to ensure that the compatibility with `master`. Thank you!

**Comment by DanTup** [comment]
@BanzaiMan do you have an approximate idea of when someone might be able to look at this to put live? I appreciate you're all busy - I have a PR out adding Windows Dart support to another project that was rather messy (manually downloading the SDK, and having to use `language: c` to have it work the same across all platforms!) and I'm trying to decide whether to land it and revert it later, or just wait for this and avoid the complication.

Thanks!

**Comment by mit-mit** [comment]
@BanzaiMan would love an update

**Comment by BanzaiMan** [comment]
@DanTup @mit-mit If you believe this is ready for deployment in production, I can take care of that, but it won't be merged until we receive a confirmation from GitHub (on a small issue). Sorry about the delay!

**Comment by mit-mit** [comment]
Yes, this is ready, please deploy

**Comment by DanTup** [comment]
@BanzaiMan Thanks! 🎉

**Pull request merged**